File: backend/app/services/text_encoding.py
"""Detección UTF-8 / Windows-1252 y reparación de mojibake en textos de importación."""

from __future__ import annotations

import re
from typing import Optional

_SPANISH = re.compile(r"[áéíóúñÁÉÍÓÚÑüÜ]")
_MOJIBAKE = re.compile(r"[\ufffd]|Ã[\x80-\xbf]|Ã.")
_CTRL = re.compile(r"[\u0080-\u009f]")
_PLUGIN_ID_RE = re.compile(r"Plugin ID:\s*(\d+)", re.IGNORECASE)


def spanish_char_count(text: str) -> int:
    return len(_SPANISH.findall(text or ""))


def extract_nessus_plugin_id(raw: Optional[str]) -> Optional[str]:
    if not raw:
        return None
    match = _PLUGIN_ID_RE.search(raw)
    return match.group(1) if match else None


def _score_text(text: str) -> int:
    score = text.count("\ufffd") * 80
    score += len(_MOJIBAKE.findall(text)) * 4
    score += len(_CTRL.findall(text)) * 2
    score -= spanish_char_count(text) * 2
    return score


def decode_bytes_smart(data: bytes) -> str:
    """Elige la mejor codificación entre UTF-8 y Windows-1252 / Latin-1 (sin decodificar UTF-8 permisivo)."""
    if not data:
        return ""
    best: tuple[int, str] | None = None
    for enc in ("utf-8-sig", "utf-8", "cp1252", "latin-1"):
        try:
            text = data.decode(enc)
        except UnicodeDecodeError:
            continue
        score = _score_text(text)
        if best is None or score < best[0]:
            best = (score, text)
    if best is not None:
        return best[1]
    return data.decode("cp1252", errors="replace")


def fix_text_encoding(text: Optional[str]) -> Optional[str]:
    """Repara mojibake (UTF-8 leído como Latin-1/CP1252) y caracteres de reemplazo."""
    if text is None:
        return None
    if not text.strip():
        return text

    candidates = [text]

    if "\ufffd" in text or "Ã" in text or _CTRL.search(text):
        for enc in ("latin-1", "cp1252"):
            try:
                fixed = text.encode(enc, errors="ignore").decode("utf-8")
                if fixed and fixed != text:
                    candidates.append(fixed)
            except (UnicodeDecodeError, UnicodeEncodeError):
                pass
        try:
            fixed2 = text.encode("utf-8", errors="ignore").decode("cp1252")
            if fixed2 and fixed2 != text:
                candidates.append(fixed2)
        except (UnicodeDecodeError, UnicodeEncodeError):
            pass

    for enc in ("latin-1", "cp1252"):
        try:
            mojibake_fix = text.encode(enc).decode("utf-8")
            if mojibake_fix != text:
                candidates.append(mojibake_fix)
        except (UnicodeDecodeError, UnicodeEncodeError):
            pass

    return min(candidates, key=_score_text)


def fix_optional_str(value: Optional[str]) -> Optional[str]:
    if value is None:
        return None
    return fix_text_encoding(value) or value

File: src/lib/text-encoding.ts
/** Detección UTF-8 vs Windows-1252 / ISO-8859-1 para CSV en español. */

const SPANISH = /[áéíóúñÁÉÍÓÚÑüÜ]/g;
const MOJIBAKE = /[\uFFFD]|Ã./g;

export type CsvEncoding = 'auto' | 'utf-8' | 'cp1252' | 'latin-1';

export function spanishCharCount(text: string): number {
  return (text.match(SPANISH) || []).length;
}

function scoreText(text: string): number {
  let score = (text.match(/\uFFFD/g) || []).length * 200;
  score += (text.match(MOJIBAKE) || []).length * 8;
  score -= spanishCharCount(text) * 3;
  return score;
}

const DECODER_LABEL: Record<Exclude<CsvEncoding, 'auto'>, string> = {
  'utf-8': 'utf-8',
  cp1252: 'windows-1252',
  'latin-1': 'iso-8859-1',
};

export function decodeCsvBytes(data: Uint8Array, encoding: CsvEncoding = 'auto'): string {
  if (encoding !== 'auto') {
    const label = DECODER_LABEL[encoding];
    try {
      return new TextDecoder(label, { fatal: true }).decode(data);
    } catch {
      return new TextDecoder(label).decode(data);
    }
  }

  const tryEncodings = ['utf-8', 'windows-1252', 'iso-8859-1'] as const;
  let best: { score: number; text: string } | null = null;

  for (const enc of tryEncodings) {
    try {
      const text = new TextDecoder(enc, { fatal: true }).decode(data);
      const s = scoreText(text);
      if (!best || s < best.score) best = { score: s, text };
    } catch {
      /* esta codificación no aplica a estos bytes */
    }
  }

  if (best && !best.text.includes('\uFFFD')) return best.text;

  const cp1252 = new TextDecoder('windows-1252').decode(data);
  if (!best || scoreText(cp1252) < best.score) return cp1252;
  return best.text;
}

export function fixTextEncoding(text: string | null | undefined): string {
  if (!text) return text ?? '';
  const trimmed = text.trim();
  if (!trimmed) return text;

  const candidates = [text];

  if (text.includes('Ã') || text.includes('\uFFFD')) {
    for (const enc of ['iso-8859-1', 'windows-1252'] as const) {
      try {
        const bytes = new Uint8Array([...text].map((c) => c.charCodeAt(0) & 0xff));
        const fixed = new TextDecoder('utf-8').decode(bytes);
        if (fixed && fixed !== text) candidates.push(fixed);
      } catch {
        /* ignore */
      }
    }
  }

  for (const enc of ['iso-8859-1', 'windows-1252'] as const) {
    try {
      const fixed = new TextDecoder('utf-8').decode(
        new TextEncoder().encode([...text].map((c) => String.fromCharCode(c.charCodeAt(0) & 0xff)).join(''))
      );
      if (fixed !== text) candidates.push(fixed);
    } catch {
      try {
        const latin = [...text].map((c) => c.charCodeAt(0) & 0xff);
        const fixed = new TextDecoder('utf-8').decode(new Uint8Array(latin));
        if (fixed !== text) candidates.push(fixed);
      } catch {
        /* ignore */
      }
    }
  }

  try {
    const mojibakeFix = new TextDecoder('utf-8').decode(
      new Uint8Array([...text].map((c) => c.charCodeAt(0) & 0xff))
    );
    if (mojibakeFix !== text) candidates.push(mojibakeFix);
  } catch {
    /* ignore */
  }

  return candidates.reduce((a, b) => (scoreText(a) <= scoreText(b) ? a : b));
}

export function repairFindingDisplayText<T extends { titulo?: string | null; descripcion?: string | null }>(
  finding: T
): T {
  return {
    ...finding,
    titulo: fixTextEncoding(finding.titulo),
    descripcion: finding.descripcion ? fixTextEncoding(finding.descripcion) : finding.descripcion,
  };
}
